### PR TITLE
[Feature] [56673] Handle Load Image - [56674] Create List Film Screen

### PR DESCRIPTION
**Pull request by ghost**
## Related Tickets
- [#Task 56673: Handle Load Image](https://edu-redmine.sun-asterisk.vn/issues/56673)
- - [#Task 56674: Create List Film Screen](https://edu-redmine.sun-asterisk.vn/issues/56674)
## WHAT
- Handle load image without lag
- Create UI for List Film Screen
- Handle Logic when click "View All"
## Evidence (Screenshot or Video)
https://user-images.githubusercontent.com/114736720/203264801-067721ec-c760-41a5-9285-2c242ce552a8.mov
## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/Zts4P8 |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Message | Sun* commit message | Template: [TASK/FREATURE/BUG/SUPPORT][RedmineID] + message |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

**Review by tunghs-1253 — APPROVED** [review]

File: MovieDB/Utils/CustomImageView.swift [hunk 1]
```diff
@@ -10,24 +10,33 @@ import UIKit
...
-            switch result {
-            case .success(let data):
-                DispatchQueue.main.async {
-                    if self.imageUrlString == url {
-                        self.image = UIImage(data: data)
+
+        if let imageFromCache = imageCache.object(forKey: urlString as NSString) {
+            self.image = imageFromCache
+            return
+        }
```
Comment (ghost) [inline MovieDB/Utils/CustomImageView.swift:21]: use `guard let`

File: MovieDB/Extension/DispatchQueue+.swift [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    static func background(delay: Double = 0.0, background: (() -> Void)? = nil, completion: (() -> Void)? = nil) {
+        DispatchQueue.global(qos: .background).async {
+            background?()
+            if let completion = completion {
+                DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: {
+                    completion()
+                })
+            }
+        }
+    }
```
Comment (ghost) [inline MovieDB/Extension/DispatchQueue+.swift:20]: func này dùng làm gì vậy e

File: MovieDB/ViewController/ListFilmVC/ListFilmViewController.swift [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    @IBOutlet private weak var categoryTitleLabel: UILabel!
+    
+    private var filmSection = FilmSection()
+    private var category: CategoryPath?
+    private var genres = [Genre]()
+    private var genresName = [String]()
+    private let filmRepository = FilmRepository()
+    private let delayRunner = DelayedRunner.initWithDuration(seconds: 0.5)
+    private var listFavoriteFilmId = [Int]()
+    static var identifier = "ListFilmViewController"
```
Comment (ghost) [inline MovieDB/ViewController/ListFilmVC/ListFilmViewController.swift:22]:
```suggestion
    static let identifier = "ListFilmViewController"
```

File: MovieDB/ViewController/ListFilmVC/ListFilmViewController.swift [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        }
+        if isFavorited {
+            CoreDataManager.shared.deleteFilmFromCoreData(filmId: filmId) { [weak self] error in
+                guard let self = self else { return }
+                if let error = error {
+                    DispatchQueue.main.async {
+                        self.showAlert(title: "ERROR", messageError: error.localizedDescription)
+                    }
+                    return
+                }
```
Comment (ghost) [inline MovieDB/ViewController/ListFilmVC/ListFilmViewController.swift:164]: use `guard let`

File: MovieDB/ViewController/ListFilmVC/ListFilmViewController.swift [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            }
+        } else {
+            CoreDataManager.shared.addFilmToCoreData(filmInfo: film) { [weak self] error in
+                guard let self = self else { return }
+                if let error = error {
+                    DispatchQueue.main.async {
+                        self.showAlert(title: "ERROR", messageError: error.localizedDescription)
+                    }
+                    return
+                }
```
Comment (ghost) [inline MovieDB/ViewController/ListFilmVC/ListFilmViewController.swift:178]: as above

File: MovieDB/Extension/DispatchQueue+.swift (hunk 2 above)
Comment (ghost) [reply MovieDB/Extension/DispatchQueue+.swift:20]: @phonglh-2668 E dùng func để load image ở cell ạ. Nếu không có hàm này thì lúc scroll collectionview nó bị giật giật ạ.

**Pull request merged**